File: packages/core-dotnet/AuthData.cs
using System.Security.Cryptography;
using System.Text;

namespace OpenPasskey.Core;

internal class ParsedAuthData
{
    public required byte[] RpIdHash { get; set; }
    public required byte Flags { get; set; }
    public required uint SignCount { get; set; }
    public byte[]? CredentialId { get; set; }
    public byte[]? CredentialKey { get; set; }
}

internal static class AuthData
{
    private const int MinAuthDataLen = 37;

    public static ParsedAuthData Parse(byte[] authData, bool expectCredData)
    {
        if (authData.Length < MinAuthDataLen)
            throw new Exception("authenticator_data_too_short");

        byte[] rpIdHash = new byte[32];
        Array.Copy(authData, 0, rpIdHash, 0, 32);

        byte flags = authData[32];

        uint signCount = (uint)(
            (authData[33] << 24) |
            (authData[34] << 16) |
            (authData[35] << 8) |
            authData[36]
        );

        var result = new ParsedAuthData
        {
            RpIdHash = rpIdHash,
            Flags = flags,
            SignCount = signCount
        };

        if (expectCredData)
        {
            if ((flags & 0x40) == 0)
                throw new Exception("no_attested_credential_data");

            if (authData.Length < 55)
                throw new Exception("authenticator_data_too_short");

            int offset = 37;
            // AAGUID: 16 bytes
            offset += 16;
            // Credential ID length: 2 bytes big-endian
            int credIdLen = (authData[offset] << 8) | authData[offset + 1];
            offset += 2;

            if (authData.Length < offset + credIdLen)
                throw new Exception("authenticator_data_too_short");

            result.CredentialId = new byte[credIdLen];
            Array.Copy(authData, offset, result.CredentialId, 0, credIdLen);
            offset += credIdLen;

            result.CredentialKey = new byte[authData.Length - offset];
            Array.Copy(authData, offset, result.CredentialKey, 0, authData.Length - offset);
        }

        return result;
    }

    public static void VerifyRpIdHash(byte[] authDataRpIdHash, string rpId)
    {
        byte[] expected = SHA256.HashData(Encoding.UTF8.GetBytes(rpId));
        if (!CryptographicOperations.FixedTimeEquals(authDataRpIdHash, expected))
            throw new WebAuthnException("rp_id_mismatch");
    }
}

File: packages/core-java/src/main/java/com/openpasskey/core/AuthData.java
package com.openpasskey.core;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

/**
 * Authenticator data parsing and RP ID hash verification.
 */
public final class AuthData {
    private AuthData() {}

    private static final ObjectMapper CBOR_MAPPER = new ObjectMapper(new CBORFactory());

    /** Parsed authenticator data fields. */
    public static class Parsed {
        public final byte[] rpIdHash;     // 32 bytes
        public final int flags;           // 1 byte
        public final int signCount;       // 4 bytes big-endian
        public final byte[] credentialId; // variable length (only if AT flag set)
        public final byte[] coseKeyBytes; // remaining bytes (only if AT flag set)

        public Parsed(byte[] rpIdHash, int flags, int signCount,
                      byte[] credentialId, byte[] coseKeyBytes) {
            this.rpIdHash = rpIdHash;
            this.flags = flags;
            this.signCount = signCount;
            this.credentialId = credentialId;
            this.coseKeyBytes = coseKeyBytes;
        }
    }

    /**
     * Parse authenticator data bytes.
     * @param authData raw authenticator data
     * @param expectAttestedCredential if true, expect AT flag and parse credential data
     */
    public static Parsed parse(byte[] authData, boolean expectAttestedCredential) throws WebAuthnException {
        if (authData.length < 37) {
            throw new WebAuthnException("invalid_authenticator_data", "Authenticator data too short");
        }

        byte[] rpIdHash = Arrays.copyOfRange(authData, 0, 32);
        int flags = authData[32] & 0xFF;
        int signCount = ByteBuffer.wrap(authData, 33, 4).getInt();

        byte[] credentialId = null;
        byte[] coseKeyBytes = null;

        if ((flags & 0x40) != 0) {
            // AT flag set: parse attested credential data
            if (authData.length < 55) {
                throw new WebAuthnException("invalid_authenticator_data",
                        "Authenticator data too short for attested credential");
            }
            // Skip AAGUID (16 bytes at offset 37)
            int credIdLen = ((authData[53] & 0xFF) << 8) | (authData[54] & 0xFF);
            int credIdEnd = 55 + credIdLen;
            if (authData.length < credIdEnd) {
                throw new WebAuthnException("invalid_authenticator_data",
                        "Authenticator data too short for credential ID");
            }
            credentialId = Arrays.copyOfRange(authData, 55, credIdEnd);
            coseKeyBytes = readFirstCborItem(authData, credIdEnd);
        }

        return new Parsed(rpIdHash, flags, signCount, credentialId, coseKeyBytes);
    }

    private static byte[] readFirstCborItem(byte[] authData, int offset) throws WebAuthnException {
        byte[] remaining = Arrays.copyOfRange(authData, offset, authData.length);
        try (JsonParser parser = CBOR_MAPPER.getFactory().createParser(remaining)) {
            CBOR_MAPPER.readValue(parser, Object.class);
            long bytesRead = parser.currentLocation().getByteOffset();
            if (bytesRead <= 0 || bytesRead > remaining.length) {
                throw new WebAuthnException("invalid_authenticator_data",
                        "Invalid credential public key length");
            }
            return Arrays.copyOf(remaining, (int) bytesRead);
        } catch (IOException e) {
            throw new WebAuthnException("invalid_authenticator_data",
                    "Failed to decode credential public key CBOR");
        }
    }

    /**
     * Verify that rpIdHash matches SHA-256(rpId).
     */
    public static void verifyRpIdHash(byte[] rpIdHash, String rpId) throws WebAuthnException {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] expected = md.digest(rpId.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            if (!MessageDigest.isEqual(rpIdHash, expected)) {
                throw new WebAuthnException("rp_id_mismatch",
                        "RP ID hash does not match expected value");
            }
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    /**
     * Compute SHA-256 hash of the given data.
     */
    public static byte[] sha256(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return md.digest(data);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }
}
